core/crypto/rng_fortuna.c

```diff
@@ -0,0 +1,518 @@
+	if (res)
+		return res;
+	if (blen % BLOCK_SIZE) {
+		uint8_t block[BLOCK_SIZE];
+		uint8_t *b = (uint8_t *)buf + ROUNDDOWN(blen, BLOCK_SIZE);
+
+		res = generate_blocks(block, 1);
+		if (res)
+			return res;
+		memcpy(b, block, blen % BLOCK_SIZE);
+	}
+
+	return TEE_SUCCESS;
+}
+
+#ifdef CFG_SECURE_TIME_SOURCE_REE
+static bool reseed_rate_limiting(void)
+{
+	/*
+	 * There's no point in checking REE time for reseed rate limiting,
+	 * and also it makes it less complicated if we can avoid doing RPC
+	 * here.
+	 */
+	return false;
+}
+#else
+static bool reseed_rate_limiting(void)
+{
+	TEE_Result res;
+	TEE_Time time;
+	const TEE_Time time_100ms = { 0, 100 };
+
+	res = tee_time_get_sys_time(&time);
+	/*
+	 * Failure to read time must result in allowing reseed or we could
+	 * block reseeding forever.
+	 */
+	if (res)
+		return false;
+
+	if (TEE_TIME_LT(time, state.next_reseed_time))
+		return true;
+
+	/* Time to reseed, calculate next time reseed is OK */
+	TEE_TIME_ADD(time, time_100ms, state.next_reseed_time);
+	return false;
+}
+#endif
+
+static TEE_Result restart_pool(void *pool_ctx, uint8_t pool_digest[KEY_SIZE])
+{
+	TEE_Result res = hash_final(pool_ctx, pool_digest);
+
+	if (res)
+		return res;
+
+	res = hash_init(pool_ctx);
+	if (res)
+		return res;
+
+	/*
+	 * Restart the pool with the digest of the old pool. This is an
+	 * extension to Fortuna. In the original Fortuna all pools was
+	 * restarted from scratch. This extension is one more defense
+	 * against spamming of the pools with known data which could lead
+	 * to the spammer knowing the state of the pools.
+	 *
+	 * This extra precaution could be useful since OP-TEE sometimes
+	 * have very few sources of good entropy and at the same time has
+	 * sources that could quite easily be predicted by an attacker.
+	 */
+	return hash_update(pool_ctx, pool_digest, KEY_SIZE);
+}
+
+static TEE_Result maybe_reseed(void)
+{
+	TEE_Result res;
+	size_t n;
+	uint8_t pool_digest[KEY_SIZE];
+
+	if (state.pool0_length < MIN_POOL_SIZE)
+		return TEE_SUCCESS;
+
+	if (reseed_rate_limiting())
+		return TEE_SUCCESS;
+
+	state.reseed_count++;
+
+	res = hash_init(state.reseed_ctx);
+	if (res)
+		return res;
+
+	for (n = 0;
+	     n < NUM_POOLS && !((state.reseed_count >> n) & 1); n++) {
+		res = restart_pool(state.pool_ctx[n], pool_digest);
+		if (res)
+			return res;
+		if (!n)
+			state.pool0_length = 0;
+
+		res = hash_update(state.reseed_ctx, pool_digest, KEY_SIZE);
+		if (res)
+			return res;
+	}
+	res = hash_final(state.reseed_ctx, pool_digest);
+	if (res)
+		return res;
+
+	crypto_cipher_final(state.ctx, CIPHER_ALGO);
+	res = crypto_cipher_init(state.ctx, CIPHER_ALGO, TEE_MODE_ENCRYPT,
+				 pool_digest, KEY_SIZE, NULL, 0, NULL, 0);
+	if (res)
+		return res;
+	inc_counter(state.counter);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result fortuna_read(void *buf, size_t blen)
+{
+	TEE_Result res;
+
+	if (!state.ctx)
+		return TEE_ERROR_BAD_STATE;
+
+	mutex_lock(&state_mu);
```

Comment:
So. this function and therefore `crypto_rng_read()` cannot be called before normal world RPC is ready, right? Then what about `tee_pager_generate_authenc_key()`?

Reply:
Since there's only one thread active during early boot there cannot be any contention on the mutex so `mutex_lock()` will always succeed directly without a need for RPC.

Reply:
However, long-term I'm planning to delay the `tee_pager_generate_authenc_key()` until after normal world has supplied a previously saved seed. Exactly how that's to be done is still an open question.